src/System.Private.ServiceModel/src/System/ServiceModel/Channels/SecurityChannelFactory.cs

```diff
@@ -0,0 +1,843 @@
+            : this(securityCapabilities, context, channelBuilder, innerChannelFactory)
+        {
+            _securityProtocolFactory = protocolFactory;
+        }
+
+        SecurityChannelFactory(ISecurityCapabilities securityCapabilities, BindingContext context, ChannelBuilder channelBuilder, IChannelFactory innerChannelFactory)
+            : base(context.Binding, innerChannelFactory)
+        {
+            _channelBuilder = channelBuilder;
+            _messageVersion = context.Binding.MessageVersion;
+            _securityCapabilities = securityCapabilities;
+        }
+
+        // used by internal test code
+        internal SecurityChannelFactory(Binding binding, SecurityProtocolFactory protocolFactory, IChannelFactory innerChannelFactory)
+            : base(binding, innerChannelFactory)
+        {
+            _securityProtocolFactory = protocolFactory;
+        }
+
+        public ChannelBuilder ChannelBuilder
+        {
+            get
+            {
+                return _channelBuilder;
+            }
+        }
+
+        public SecurityProtocolFactory SecurityProtocolFactory
+        {
+            get
+            {
+                return _securityProtocolFactory;
+            }
+        }
+
+        public SecuritySessionClientSettings<TChannel> SessionClientSettings
+        {
+            get
+            {
+                Fx.Assert(SessionMode == true, "SessionClientSettings can only be used if SessionMode == true");
+                return _sessionClientSettings;
+            }
+        }
+
+        public bool SessionMode
+        {
+            get
+            {
+                return _sessionMode;
+            }
+        }
+
+        bool SupportsDuplex
+        {
+            get
+            {
+                ThrowIfProtocolFactoryNotSet();
+                return _securityProtocolFactory.SupportsDuplex;
+            }
+        }
+
+        bool SupportsRequestReply
+        {
+            get
+            {
+                ThrowIfProtocolFactoryNotSet();
+                return _securityProtocolFactory.SupportsRequestReply;
+            }
+        }
+
+        public MessageVersion MessageVersion
+        {
+            get
+            {
+                return _messageVersion;
+            }
+        }
+
+        void CloseProtocolFactory(bool aborted, TimeSpan timeout)
+        {
+            if (_securityProtocolFactory != null && !SessionMode)
+            {
+                _securityProtocolFactory.Close(aborted, timeout);
+                _securityProtocolFactory = null;
+            }
+        }
+
+        async Task CloseProtocolFactoryAsync(TimeSpan timeout)
+        {
+            if (_securityProtocolFactory != null && !SessionMode)
+            {
+                await _securityProtocolFactory.CloseAsync(timeout);
+                _securityProtocolFactory = null;
+            }
+        }
```

Comment:
I'm guessing this code was as before, but we probably should snapshot _securityProtocolFactory before .Close-ing - there's a potential null ref here. Also for the sync path. 
Additionally, consider consolidating so that Close() just calls CloseAsync().GetAwaiter().GetResult() so we don't have to maintain two paths